File: wall.cpp


#include "wall.h"
#include <QtMath>

bool Wall::isZero(double d) {
    return qAbs(d) < 1e-6;
}

Wall::Wall(const Pos &begin, const Pos end) {
    begin_ = Pos(begin.getX(), begin.getY());
    end_ = Pos(end.getX(), end.getY());
}

Wall::Wall(double x1, double y1, double x2, double y2) {
    begin_ = Pos(x1, y1);
    end_ = Pos(x2, y2);
}

Pos Wall::crossPoint(const Pos &pos) {
    // 如果是 x 轴平行线或者 y 轴平行线
    if (isZero(begin_.getX() - end_.getX())) {
        // 直线方程 x = begin_.getX();
        return Pos(begin_.getX(), pos.getY());
    } else if (isZero(begin_.getY() - end_.getY())) {
        // 直线方程 y = begin_.getY();
        return Pos(pos.getX(), end_.getY());
    } else {
        double aW = (end_.getY() - begin_.getY()) / (end_.getX() - begin_.getX())
                , bW = begin_.getY() - aW * begin_.getX(), aV = (-1) / aW
                        , bV = pos.getY() - aV * pos.getX();
        return Pos((bV - bW) / (aW - aV), aW * ((bV - bW) / (aW - aV)) + bW);
    }
}

bool Wall::isIn(const Pos &pos, double r) {
    Pos metaDir = end_.normalize(begin_).multiply(r / 2.0);
    Pos xBegin = begin_.subtract(metaDir);
    Pos xEnd = end_.add(metaDir);
    // 判断是否在直线内
    return pos.getX() >= qMin(xBegin.getX(), xEnd.getX())
    && pos.getX() <= qMax(xBegin.getX(), xEnd.getX())
    && pos.getY() >= qMin(xBegin.getY(), xEnd.getY())
    && pos.getY() <= qMax(xBegin.getY(), xEnd.getY());
}

double Wall::distanceTo(const Pos &pos) {
    // 点到直线的距离
    Pos tmpPos = crossPoint(pos);
    return tmpPos.distanceTo(pos);
}

Pos Wall::getBegin() {
    return Pos(begin_.getX(), begin_.getY());
}

Pos Wall::getEnd() {
    return Pos(end_.getX(), end_.getY());
}

File: Wall.cs
using System;

namespace OSM
{
    public class Wall
    {
        private Vector begin, end;

        public Wall(Vector p, Vector q)
        {
            begin = p;
            end = q;
        }

        public Wall(double x, double y, double p, double q)
        {
            begin = new Vector(x, y);
            end = new Vector(p, q);
        }
        
        // 计算点在直线上的交点
        public Vector CrossPoint(Vector vector)
        {
            // 直线方程 y = a * x + b
            // 考虑 斜率为 0  或者 无穷大
            if (CustomizeUtil.IsZero(begin.GetX() - end.GetX())) 
            {
                // 直线的方程 : x = begin.GetX();
                return new Vector(begin.GetX(), vector.GetY());
            } 
            else if (CustomizeUtil.IsZero(begin.GetY() - end.GetY())) 
            {
                // 直线 方程: y = begin.GetY();
                return new Vector(vector.GetX(), begin.GetY());
            }
            else
            {
                double aW = (end.GetY() - begin.GetY()) / (end.GetX() - begin.GetX())
                    , bW = begin.GetY() - aW * begin.GetX()
                    , aV =  (-1) / aW
                    , bV = vector.GetY() - aV * vector.GetX();
                // 交点
                return new Vector((bV - bW) / (aW - aV), aW * ((bV - bW) / (aW - aV)) + bW);
            }
        }
        
        public bool IsIn(Vector vector, double r) 
        {
            Vector metaDir = end.Normalize(begin).Multiply(r / 2.0);
            Vector xBegin = begin.NewSubtract(metaDir), xEnd = end.NewAdd(metaDir);

            return vector.GetX() >= Math.Min(xBegin.GetX(), xEnd.GetX())
                   && vector.GetX() <= Math.Max(xBegin.GetX(), xEnd.GetX())
                   && vector.GetY() >= Math.Min(xBegin.GetY(), xEnd.GetY())
                   && vector.GetY() <= Math.Max(xBegin.GetY(), xEnd.GetY());

        }
        
        public double DistanceTo(Vector vector) 
        {
            // 点到线的距离
            return vector.DistanceTo(CrossPoint(vector));
        }
        
        
        public Vector GetBegin() 
        {
            return begin;
        }

        public Vector GetEnd() 
        {
            return end;
        }

        
    }
}